video/out/opengl/hwdec_ios.m

```diff
@@ -0,0 +1,209 @@
+        NULL,
+        [EAGLContext currentContext],
+        NULL,
+        &p->gl_texture_cache);
+
+    if (err != noErr) {
+        MP_ERR(hw, "Failure in CVOpenGLESTextureCacheCreate: %d\n", err);
+        return -1;
+    }
+
+    p->hwctx = (struct mp_hwdec_ctx){
+        .type = HWDEC_VIDEOTOOLBOX,
+        .download_image = mp_vt_download_image,
+        .ctx = &p->hwctx,
+    };
+    hwdec_devices_add(hw->devs, &p->hwctx);
+
+    return 0;
+}
+
+static int reinit(struct gl_hwdec *hw, struct mp_image_params *params)
+{
+    struct priv *p = hw->priv;
+    assert(params->imgfmt == hw->driver->imgfmt);
+
+    if (!params->hw_subfmt) {
+        MP_ERR(hw, "Unsupported CVPixelBuffer format.\n");
+        return -1;
+    }
+
+    if (!gl_get_imgfmt_desc(hw->gl, params->hw_subfmt, &p->desc)) {
+        MP_ERR(hw, "Unsupported texture format.\n");
+        return -1;
+    }
+
+    params->imgfmt = params->hw_subfmt;
+    params->hw_subfmt = 0;
+    return 0;
+}
+
+static void cleanup_textures(struct gl_hwdec *hw)
+{
+    struct priv *p = hw->priv;
+    int i;
+
+    for (i = 0; i < MP_MAX_PLANES; i++) {
+        if (p->gl_planes[i]) {
+            CFRelease(p->gl_planes[i]);
+            p->gl_planes[i] = NULL;
+        }
+    }
+
+    CVOpenGLESTextureCacheFlush(p->gl_texture_cache, 0);
+}
+
+static int map_frame(struct gl_hwdec *hw, struct mp_image *hw_image,
+                     struct gl_hwdec_frame *out_frame)
+{
+    struct priv *p = hw->priv;
+    GL *gl = hw->gl;
+
+    CVPixelBufferRelease(p->pbuf);
+    p->pbuf = (CVPixelBufferRef)hw_image->planes[3];
+    CVPixelBufferRetain(p->pbuf);
+
+    const bool planar = CVPixelBufferIsPlanar(p->pbuf);
+    const int planes  = CVPixelBufferGetPlaneCount(p->pbuf);
+    assert((planar && planes == p->desc.num_planes) || p->desc.num_planes == 1);
+
+    cleanup_textures(hw);
+
+    for (int i = 0; i < p->desc.num_planes; i++) {
+        const struct gl_format *fmt = p->desc.planes[i];
+        GLenum format = fmt->format;
+        GLenum internal_format = fmt->internal_format;
+
+        if (hw->gl->es >= 300) {
+            if (format == GL_RED) {
+                format = internal_format = GL_LUMINANCE;
+            } else if (format == GL_RG) {
+                format = internal_format = GL_LUMINANCE_ALPHA;
+            }
+        }
+
+        CVReturn err = CVOpenGLESTextureCacheCreateTextureFromImage(
+            kCFAllocatorDefault,
+            p->gl_texture_cache,
+            p->pbuf,
+            NULL,
+            GL_TEXTURE_2D,
+            internal_format,
+            CVPixelBufferGetWidthOfPlane(p->pbuf, i),
+            CVPixelBufferGetHeightOfPlane(p->pbuf, i),
+            format,
+            fmt->type,
+            i,
+            &p->gl_planes[i]);
+
+        if (err != noErr) {
+            MP_ERR(hw, "error creating texture for plane %d: %d\n", i, err);
+            return -1;
+        }
+
+        gl->BindTexture(GL_TEXTURE_2D, CVOpenGLESTextureGetName(p->gl_planes[i]));
```

Comment:
It should be unbound either after the texparameters, or after the loop (as long as nothing is bound when the function is left).